lib/AST/ASTPrinter.cpp

```diff
@@ -3586,19 +3589,81 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
     return T->hasSimpleTypeRepr();
   }
 
+  /// Add all Clang modules that have been imported into `Mod` to
+  /// `clangModules`.
+  ///
+  /// This also takes into account any modules that are imported transitively
+  /// through public (`@_exported`) imports.
+  /// `clangModules` is a map from each imported Clang module to the
+  /// corresponding Swift module.
+  static void getImportedClangModules(
```

Comment:
> WDYT about `computeVisibleClangModules`? That would make a more obvious connection with `getVisibleClangModules` and emphasize the get vs. compute distinction.

I consciously made this distinction because this is a lower-level function that, on its own, doesn't really guarantee that you'll get exactly the visible modules; that only happens if you pass the right value for `filter`.

I think this concern may go away though once I rewrite the function as you suggest below.

> Could you change `getImportedClangModules` to return the may by value, and to keep a data recursion queue, and to keep a set of already-visited modules? Module interface printing can get called from performance-sensitive contexts in the IDE, so I'd rather not take chances with an exponential algorithm.

That makes sense. Not done yet, but I wanted to respond. I'll put a comment on this thread when I've made this change.

Reply:
> Could you change `getImportedClangModules` to return the may by value, and to keep a data recursion queue, and to keep a set of already-visited modules? Module interface printing can get called from performance-sensitive contexts in the IDE, so I'd rather not take chances with an exponential algorithm.

Done.